### PR TITLE
[WIP][tool] Add scripts to generate test data for WPT WebNN operations tests

### PR DESCRIPTION
**Pull request by BruceDai**
@huningxin @fdwr I added these scripts to generate `float64` inputs test data and `float32` expected data for WebNN concat tests of WPT, PTAL, thanks.

### CODE REVIEW

File: test/tool/gen-concat.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        if (toSaveDataDict['expectedData'][expectedDataCategory] === undefined) {
+          const precisionType = test.type;
+          const inputShapes = test.inputs.shape;
+          const inputDataCategory = test.inputs.data;
+          const feedData = toSaveDataDict['inputsData'][inputDataCategory];
+          const inputShapeValues = [];
+          let pos = 0;
+          for (const shape of inputShapes) {
+            const size = sizeOfShape(shape);
+            const precisionData =
```
Comment (BruceDai) [inline test/tool/gen-concat.js:68]: This Line68 is to convert origin `float64` data to `float32` before invoking the implementation of `concat` operation, and below Line74 is to convert computed result to be `float32` data. @huningxin PTAL, thank.

File: test/tool/resources/concat.json [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+          },
+          "expected": {
+            "shape": [1, 2, 1, 1, 12],
+            "data": "float32Axis4Two5D"
+          },
+          "type": "float32"
+        }
+      ]
+    }
+  },
```
Comment (BruceDai) [inline test/tool/resources/concat.json:263]: Above `tests` part is reused the [`tests` part](https://github.com/BruceDai/wpt/blob/add_webnn_concat_tests/webnn/resources/test_data/concat.json#L2) of test_data file for WebNN `concat` tests on WPT.

**Comment by BruceDai** [comment]
Link https://github.com/web-platform-tests/wpt/pull/36782 .

File: test/tool/gen-concat.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      const axis = test.axis;
+      const precisionType = test.type;
+      const inputShapes = test.inputs.shape;
+      const inputDataCategory = test.inputs.data;
+      const feedData = toSaveDataDict['inputsData'][inputDataCategory];
+      const inputShapeValues = [];
+      let pos = 0;
+      for (const shape of inputShapes) {
+        const size = sizeOfShape(shape);
+        const precisionData =
```
Comment (BruceDai) [inline test/tool/gen-concat.js:65]: This Line65 is to convert origin float64 data to float32 before invoking the implementation of `concat` operation, and below Line71 is to convert computed result to be float32 data based on our last synced solution. @huningxin PTAL, thank.

File: test/tool/resources/concat.json [hunk 4]
```diff
@@ -0,0 +1,303 @@
+{
+  "tests": [
```
Comment (BruceDai) [inline test/tool/resources/concat.json:2]: This tests part is reused the [tests part](https://github.com/BruceDai/wpt/blob/add_webnn_concat_tests/webnn/resources/test_data/concat.json#L2) of test_data file for WPT WebNN concat tests.

File: test/tool/test_data/concat_data.json [hunk 5]
```diff
@@ -0,0 +1,370 @@
+{
+  "inputsData": {
```
Comment (BruceDai) [inline test/tool/test_data/concat_data.json:2]: These generated test data of `inputsData` and `expectedData` are used as the [data](https://github.com/BruceDai/wpt/blob/add_webnn_concat_tests/webnn/resources/test_data/concat.json#L295) of test_data file for WPT WebNN concat tests.

File: test/tool/test_data/clamp_data.json [hunk 6]
```diff
@@ -0,0 +1,344 @@
+{
+  "inputsData": {
+    "float64Data": [
```
Comment (fdwr) [inline test/tool/test_data/clamp_data.json:3]: So this float64 data was cast to float32 first (and then back to float64) before passing to the baseline reference engine? Likewise, the data is also cast to float32 in the same way when passed to the web API implementation?

We just want to be sure that both inputs see the same initial tensor. If the tests are passing, then I guess they must be cast consistently.

Comment (BruceDai) [reply test/tool/test_data/clamp_data.json:3]: > So this float64 data was cast to float32 first (and then back to float64) before passing to the baseline reference engine? Likewise, the data is also cast to float32 in the same way when passed to the web API implementation?

Yes. The float64 data was cast to float32 by same invoking `new Float32Array(float64Data)`  before passing to baseline reference engine and web API implementation.

Comment (fdwr) [reply test/tool/test_data/clamp_data.json:3]: Cool.

File: test/tool/gen-gemm.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        }
+        if (options.bTranspose !== undefined) {
+          gemmOptions.bTranspose = options.bTranspose;
+        }
+      }
+      const result = gemmCompute(a, b, gemmOptions);
+      toSaveDataDict['expectedData'][expectedDataCategory] =
+          utils.getPrecisionData(result, precisionType);
+    } else {
+      continue;
```
Comment (fdwr) [inline test/tool/gen-gemm.js:78]: Is the `continue` not redundant given the next line will reach the bottom of the `for` loop? If you want to keep it, consider inverting the `if` test, moving the continue to the top and unindenting the normal code flow (fewer branch points make it a little easier to follow):

```
if (toSaveDataDict['expectedData'][expectedDataCategory] !== undefined)
{
    // The expected data is already defined.
    continue;
}

const precisionType = test.type;
const precisionDataA = utils.getPrecisionDataFromDataDict(toSaveDataDict['inputsData'], test.inputs.a.data, precisionType);
...
```

File: test/tool/gen-concat.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  // use toSaveDataDict variable to save inputs data which are new generated or
+  // already recorded in existed test data file
+  const toSaveDataDict = utils.prepareInputsData(inputsDataInfo, savedDataFile, min, max);
+  // use toSaveDataDict variable to also save expected data genertated by concat tests
+  // with above input data
+  toSaveDataDict['expectedData'] = {};
+  const tests = jsonDict.tests;
+
+  for (const test of tests) {
+    const expectedDataCategory = test.expected.data;
```
Comment (fdwr) [inline test/tool/gen-concat.js:34]: Selecting which named entry is the source of your data feels more like a *source* than a *category*. How about `expectedDataSource`?

File: test/tool/utils.js [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+  const parentDirectory = path.dirname(saveFile);
+  if (!fs.existsSync(parentDirectory)) {
+    fs.mkdirSync(parentDirectory);
+  }
+  const jsonString = JSON.stringify(jsonDict, function(key, value) {
+    // the replacer function is looking for some typed arrays.
+    // If found, it replaces it by a trio
+    if ( value instanceof Int8Array ||
+         value instanceof Uint8Array ||
+         value instanceof Uint16Array ||
```
Comment (fdwr) [inline test/tool/utils.js:147]: Int16Array?

**Review by fdwr — APPROVED** [review]

**Review by fdwr — APPROVED** [review]

**Comment by BruceDai** [comment]
Thanks all for reviewing and sharing guidelines. 👍 👍 
Let me close this earlier several PRs, and new PR of generating baseline scripts including fp16 baseline data will submit later.

**Pull request closed without merging**